libutils/proc_manager.h

```diff
@@ -0,0 +1,123 @@
+    char *description;    /** human-friendly description of the subprocess [can be %NULL] */
+    pid_t pid;            /** PID of the subprocess                        [always >= 0] */
+    int in_fd;            /** FD of the input to the subprocess            [-1 if N/A] */
+    int out_fd;           /** FD of the output from the subprocess         [-1 if N/A] */
+    char lookup_fd;       /** which FD to use for lookup by FD             ['i', 'o' or '\0'] */
+};
+typedef struct _Subprocess Subprocess;
+
+/**
+ * Destroy a subprocess.
+ */
+void SubprocessDestroy(Subprocess *proc);
+
+typedef struct _ProcManager ProcManager;
+
+/**
+ * Function to terminate a subprocess.
+ *
+ * @param subprocess_pid PID of the subprocess to terminate
+ * @param in_fd          FD of the input to the subprocess
+ * @param out_fd         FD of the output from the subprocess
+ * @param data           custom data passed to the function through the termination functions
+ * @return               whether the termination was successful or not
+ */
+typedef bool (*ProcessTerminator) (pid_t subprocess_pid, int in_fd, int out_fd, void *data);
+
+/**
+ * Get a new ProcManager.
+ */
+ProcManager *ProcManagerNew();
+
+/**
+ * Destroy a ProcManager.
+ */
+void ProcManagerDestroy();
+
+/**
+ * Add a subprocess to a ProcManager.
+ *
+ * @param manager     the ProcManager to add the process to        [cannot be %NULL]
+ * @param id          ID of the process to add                     [string(pid) is used if %NULL]
+ * @param cmd         command used to create the subprocess        [can be %NULL]
+ * @param description human-friendly description of the subprocess [can be %NULL]
+ * @param pid         PID of the subprocess                        [required to be >=0]
+ * @param in_fd       FD of the input to the subprocess            [-1 if N/A]
+ * @param out_fd      FD of the output from the subprocess         [-1 if N/A]
+ * @param lookup_fd   which FD to use for lookup by FD             ['i', 'o' or '\0']
+ *
+ * @note Ownership of #id, #cmd and #description is transferred to the #manager.
+ */
+bool ProcManagerAddProcess(ProcManager *manager,
+                           char *id, char *cmd, char *description,
+                           pid_t pid, int in_fd, int out_fd, char lookup_fd);
+
+/**
+ * Getter functions for #ProcManager.
+ */
+Subprocess *ProcManagerGetProcessByPID(ProcManager *manager, pid_t pid);
+Subprocess *ProcManagerGetProcessById(ProcManager *manager, const char *id);
+Subprocess *ProcManagerGetProcessByFD(ProcManager *manager, int fd);
+
+/**
+ * Functions to get and remove subprocesses from a #ProcManager.
+ *
+ * @note Destroy the returned #Subprocess with SubprocessDestroy().
+ */
+Subprocess *ProcManagerPopProcessByPID(ProcManager *manager, pid_t pid);
+Subprocess *ProcManagerPopProcessById(ProcManager *manager, const char *id);
+Subprocess *ProcManagerPopProcessByFD(ProcManager *manager, int fd);
```

Comment:
Is not the term "pop" related to stacks? Could perhaps "remove" be better phrasing?

Reply:
I think *pop* is quite generic for the case where you *get* something while *removing* it at the same time.